add ephemeral resource contributor docs (#40201)
hashicorp/terraform-provider-aws: docs/add-a-new-ephemeral-resource.md

<!-- markdownlint-configure-file { "code-block-style": false } -->
# Adding a New Ephemeral Resource

> New ephemeral resources are required when AWS adds a new service, or adds new features within an existing service which would require a new ephemeral resource. Ephemeral resources produce ephemeral values and are never stored to state. Anything that produces a sensitive value can be an ephemeral resource, but some are more useful than others.

Reviewer: ```suggestion
New ephemeral resources are required when AWS introduces a new service or adds features to an existing service that necessitate a new ephemeral resource. Ephemeral resources produce ephemeral values and are never stored in the state. Any resource that produces a sensitive value can be an ephemeral resource, though some are more useful than others.
```

> Each ephemeral resource should be submitted for review in isolation, pull requests containing multiple ephemeral resources and/or resources are harder to review and the maintainers will normally ask for them to be broken apart.

Reviewer: ```suggestion
Each ephemeral resource should be submitted for review individually. Pull requests containing multiple ephemeral resources or other resources are more difficult to review, and maintainers will typically request that they be split into separate submissions.
```

## Prerequisites

> If this is the first addition of a ephemeral resource for a new service, please ensure the Service Client for the new service has been added and merged. See [Adding a new Service](add-a-new-service.md) for details.

Reviewer: ```suggestion
If an ephemeral resource is the first addition for a new service, ensure that the Service Client for the service has been created and merged first. Refer to [Adding a New Service](add-a-new-service.md) for detailed instructions.
```

## Steps to Add a Ephemeral Resource

### Fork the Provider and Create a Feature Branch

> For a new ephemeral resource use a branch named `f-{datasource name}` for example: `f-ec2-vpc`. See [Raising a Pull Request](raising-a-pull-request.md) for more details.

Reviewer: ```suggestion
For a new ephemeral resource use a branch named `f-{ephemeral-resource-name}` for example: `f-kms-secret`. See [Raising a Pull Request](raising-a-pull-request.md) for more details.
```

### Create and Name the Ephemeral Resource

See the [Naming Guide](naming.md#resources-and-data-sources) for details on how to name the new ephemeral resource and the ephemeral resource file. Not following the naming standards will cause extra delay as maintainers request that you make changes.

Use the [skaff](skaff.md) provider scaffolding tool to generate new ephemeral resource and test templates using your chosen name. Doing so will ensure that any boilerplate code, structural best practices and repetitive naming are done for you and always represent our most current standards.

### Fill out the Ephemeral Resource Schema

> In the `internal/service/<service>/<service>_ephemeral.go` file you will see a `Schema` property which exists as a map of `Schema` objects. This relates the AWS API data model with the Terraform resource itself. For each property you want to make available in Terraform, you will need to add it as an attribute, and choose the correct data type.

Reviewer: ```suggestion
In the `internal/service/<service>/<service>_ephemeral.go` file, you'll find a `Schema` property, which is a map of `Schema` objects. This maps the AWS API data model to the Terraform resource. To make a property available in Terraform, add it as an attribute with the appropriate data type.
```

> Attribute names are to be specified in `snake_case` as opposed to the AWS API which is `CamelCase`.

Reviewer: ```suggestion
Define attributes using `snake_case`, instead of the `CamelCase` format used by the AWS API.
```

### Implement Open Handler

> These will map the AWS API response to the ephemeral resource schema. You will also need to handle different response types (including errors correctly). For complex attributes you will need to implement Flattener or Expander functions. The [Data Handling and Conversion Guide](data-handling-and-conversion.md) covers everything you need to know for mapping AWS API responses to Terraform State and vice-versa. The [Error Handling Guide](error-handling.md) covers everything you need to know about handling AWS API responses consistently.

Reviewer: ```suggestion
`Open` will map the AWS API response to the ephemeral resource schema. You’ll also need to handle different response types, including errors, correctly. You will typically use `Autoflex` for mapping AWS API responses to Terraform models and vice versa. The [Error Handling Guide](error-handling.md) covers best practices for handling AWS API responses consistently.
```

### Register Ephemeral Resource to the provider

> Ephemeral Resources use a self-registration process that adds them to the provider using the `@EphemeralResource()` annotation in the ephemeral resource's comments. Run `make gen` to register the ephemeral resource. This will add an entry to the `service_package_gen.go` file located in the service package folder.

Reviewer: ```suggestion
Ephemeral resources use a self-registration process that adds them to the provider via the `@EphemeralResource()` annotation in the resource's comments. To register the ephemeral resource, run `make gen`. This will generate an entry in the `service_package_gen.go` file located in the service package folder.
```

```go
package something

import (
	"context"
	
	"github.com/hashicorp/terraform-plugin-framework/ephemeral"
	"github.com/hashicorp/terraform-provider-aws/internal/framework"
)

// @EphemeralResource(name="Example")
func newEphemeralExample(_ context.Context) (ephemeral.EphemeralResourceWithConfigure, error) {
	return &ephemeralExample{}, nil
}

type ephemeralExample struct {
	framework.EphemeralResourceWithConfigure
}

func (r *ephemeralExample) Metadata(_ context.Context, request ephemeral.MetadataRequest, response *ephemeral.MetadataResponse) {
	response.TypeName = "aws_something_example"
}
```

### Write Passing Acceptance Tests

> To adequately test the ephemeral resource we will need to write a complete set of Acceptance Tests. You will need an AWS account for this which allows the provider to read to state of the associated resource. See [Writing Acceptance Tests](running-and-writing-acceptance-tests.md) for a detailed guide on how to approach these.

Reviewer: ```suggestion
To properly test the ephemeral resource, write a complete set of acceptance tests. An AWS account is required, which allows the provider to communicate with the AWS API. For a detailed guide on how to write and run these tests, refer to [Writing Acceptance Tests](running-and-writing-acceptance-tests.md).
```
